### PR TITLE
Add Goal Count Heuristic

### PR DESCRIPTION
**Pull request by williamshen-nz**
Add goal count heuristic.

Tested `python src/main.py --env painting --approach oracle --seed 0` successfully with `CFG.sesame_task_planning_heuristic == "goal_count"`. A lot of the skeleton searches time out as one might expect.

```
Planning succeeded! Found plan of length 11 after 1 skeletons, discovering 0 failures
Task 1 / 50: SOLVED
Task 2 / 50: Approach failed to solve with error: Planning timed out in skeleton search!
Planning succeeded! Found plan of length 11 after 2 skeletons, discovering 1 failures
Task 3 / 50: SOLVED
Task 4 / 50: Approach failed to solve with error: Planning timed out in skeleton search!
Planning succeeded! Found plan of length 11 after 1 skeletons, discovering 0 failures
Task 5 / 50: SOLVED
...
Task 48 / 50: Approach failed to solve with error: Planning timed out in skeleton search!
Planning succeeded! Found plan of length 12 after 2 skeletons, discovering 1 failures
Task 49 / 50: SOLVED
Task 50 / 50: Approach failed to solve with error: Planning timed out in skeleton search!
Tasks solved: 13 / 50
Average time for successes: 4.19658 seconds
Test results: defaultdict(<class 'float'>, {'num_solved': 13, 'num_total': 50, 'avg_suc_time': 4.196577365581806, 'min_skeletons_optimized': 1.0, 'max_skeletons_optimized': 2.0, 'avg_execution_failures': 0.0, 'avg_num_skeletons_optimized': 1.6153846153846154, 'avg_num_nodes_expanded': 36138.692307692305, 'avg_num_nodes_created': 135432.15384615384, 'avg_num_nsrts': 11.0, 'avg_num_preds': 13.0, 'avg_plan_length': 11.76923076923077, 'avg_num_failures_discovered': 0.6153846153846154, 'num_transitions': 0, 'learning_time': 0.0})
Wrote out test results to results/painting__oracle__0______None.pkl


Main script terminated in 443.95387 seconds

Process finished with exit code 0

```

Ran pre-checks locally and all succeeded.

___

**Some general dev notes:**

- I found some of the pylint checks a bit annoying, e.g.

  1. Cannot use the code below with `elif` and instead need to use `if`
  ```
  if x:
     return y
  elif a:
     return b
  ```
  2. Docstring for every method
  3. No TODO or FIXME comments
- Also adding a `isort` might be good for automatically sorting imports
- The repo is setup such that everything must be imported with `predicators.src`. However, this means that the repo must be cloned into a folder called `predicators` which is a bit limiting. I had to setup my Pycharm environment in a non-standard way because of this.
- Adding a script to run checks locally would make it nice. This is what I'm using: https://gist.github.com/williamshen-nz/1114d8b025506ac005c20986fc7c078c

### CODE REVIEW

File: src/utils.py [hunk 1]
```diff
@@ -1057,6 +1059,19 @@ def __call__(self, atoms: Collection[GroundAtom]) -> float:
         raise NotImplementedError("Override me!")
 
 
+class GoalCountHeuristic(_TaskPlanningHeuristic):
+    """The number of goal atoms that are not in the current state."""
+    HEURISTIC_NAME: str = "goal_count"
+
+    @functools.cached_property
+    def goal_set(self) -> Set[GroundAtom]:
```
[the remark this answers is not in the document]
Comment (williamshen-nz) [reply src/utils.py:1067]: Will do

File: tests/test_utils.py [hunk 2]
```diff
@@ -1332,37 +1332,86 @@ def test_operator_application():
     assert next_atoms == {pred1([cup1, plate1]), pred2([cup1, plate1])}
 
 
-def test_create_task_planning_heuristic():
+@pytest.mark.parametrize("heuristic_name, expected_heuristic_cls", [
```
[the remark this answers is not in the document]
Comment (williamshen-nz) [reply tests/test_utils.py:1335]: Basically an easy, cleaner and more compact way to specify lots of test cases. You specify the parameters to the test method as the first parameter of `@pytest.mark.parametrize` and then a list of test cases as the second parameter. 

Each test case will then be run by pytest as essentially a separate unit test. pytest also has a lot of other powerful things like fixtures which can make tests a lot easier to setup by reusing certain objects across tests.

[the remark this answers is not in the document]
Comment (ronuchit) [reply tests/test_utils.py:1342]: ah gotcha! sorry we are noobs 😭 it would be great to start using these more powerful features going forward

**Comment by williamshen-nz** [comment]
Thanks!

1. Regarding `elif` pattern. Agree that what pylint is saying is true. Just when refactoring and moving `return` out of the `if` statement it will require a few more extra characters :)
    - At least for me, TODOs and FIXMEs are reminders/recommendations for changes to make when I touch a certain part of the code again (not big enough to be a proper issue) but what you said makes sense. 
    - e.g. in this case, I tried adding a TODO to "refactor heuristic stuff into a `heuristics.py` class if we need to add new heuristics in the future."
3. I'll look into `isort`, will need to look into how to run it so `yapf` is happy.
4. Works fine now! 
5. The `cd predicators` was for getting it to work on my Pycharm setup. Will create a separate PR for the script and `isort/yapf` stuff later.

**Comment by williamshen-nz** [comment]
Closing this PR in favour of https://github.com/Learning-and-Intelligent-Systems/predicators/pull/497
This PR was based on my fork which messed things up

> Ah, I see. Feel free to add that as a plain old comment without a TODO! That's usually what we do in these situations. E.g. # Note: If we add new heuristics in the future, we'll want to refactor the below code into a new heuristics.py class.

Sure thing - the reason I use TODOs and FIXMEs is because Pycharm will highlight them in yellow so easier to spot

**Pull request closed without merging**